### PR TITLE
Akamai Function Worker as a Service

### PR DESCRIPTION
**Pull request by jamesm-macrometa**
This PR adds docs for integration of Akamai with Macrometa GDN.

### CODE REVIEW

**Comment by jamesm-macrometa** [comment]
@diana-macrometa I formatted this following the same pattern as Query Workers, but I can easily change it to follow our more conventional formatting if that's preferable (e.g. Create Streams)

**Comment by jamesm-macrometa** [comment]
@diana-macrometa @dlozina-macrometa @dwivedianurag Our docs on Akamai integration via GUI are ready for review.  

Do we want to include any additional concept info on Akamai and its integration with Macrometa?  If so we change the current Index into a Task, and create a new Index that contains concept stuff.

I'm still a bit unclear on the "edge worker" vs "edge function" terminology and which goes where.

File: docs/queryworkers/akamai/index.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
+---
+sidebar_position: 1
+title: Akamai in Macrometa
+---
+
+This page describes how to integrate Akamai's edge worker functionality as a Macrometa function.
+
+## Prerequisite
```
Comment (diana-macrometa) [inline docs/queryworkers/akamai/index.md]: Also need to have an account and sufficient permissions.

File: docs/queryworkers/akamai/index.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
+---
+sidebar_position: 1
+title: Akamai in Macrometa
+---
+
+This page describes how to integrate Akamai's edge worker functionality as a Macrometa function.
```
Comment (diana-macrometa) [inline docs/queryworkers/akamai/index.md]: Should "edge worker" be capitalized? Seems like we are using a proper name here.

File: docs/queryworkers/akamai/index.md (hunk 1 above)
Comment (diana-macrometa) [reply docs/queryworkers/akamai/index.md]: Probably should mention what permissions are needed. Admin? Read/write? For fabric or collection or what?

File: docs/queryworkers/akamai/view-edge-workers.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
+---
+sidebar_position: 30
+title: View Edge Workers
+---
+
+To view edge workers, click **Functions**.
+
+Macrometa displays a list of custom functions. Edge workers can be sorted by the type `Akamai`.
```
Comment (diana-macrometa) [inline docs/queryworkers/akamai/view-edge-workers.md]: Bold, not code, for most UI elements. Refer to [Google Developer Style Guide](https://developers.google.com/style/ui-elements) for specifics.

File: docs/queryworkers/akamai/index.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+## Prerequisite
+
+In Akamai, create a Property for your Macrometa GDN integration.
+
+## Add Integration
```
Comment (diana-macrometa) [inline docs/queryworkers/akamai/index.md]: Put the integration tasks in a Manage Integrations topic.

File: docs/queryworkers/akamai/generate-edge-workers.md [hunk 5]
```diff
@@ -0,0 +1,15 @@
+---
+sidebar_position: 10
+title: Generate Edge Workers
+---
+
+Generate an edge worker.
+
+1. [Log in to your Macrometa account](https://auth.paas.macrometa.io/).
+1. Click **Query Workers**.
```
Comment (diana-macrometa) [inline docs/queryworkers/akamai/generate-edge-workers.md]: Need two tasks in this topic.

Generate Edge Worker from Query Worker
Generate Edge Worker from Stream Worker

File: docs/queryworkers/akamai/index.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
+---
+sidebar_position: 1
+title: Akamai in Macrometa
+---
+
+This page describes how to integrate Akamai's edge worker functionality as a Macrometa function.
+
```
Comment (diana-macrometa) [inline docs/queryworkers/akamai/index.md:7]: The index should have more concept info as well as the information about this being a preview only. A note to that effect should be at the top of each topic as well.

Comment (diana-macrometa) [reply docs/queryworkers/akamai/index.md:7]: @dallas-macrometa or @dlozina-macrometa  might be able to help you with the concept info. What in broad strokes an Akamai edge worker is, how it extends Macrometa capabilities (JavaScript mostly), and maybe a couple use cases.

File: docs/queryworkers/akamai/index.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+## Prerequisite
+
+In Akamai, create a Property for your Macrometa GDN integration.
```
Comment (diana-macrometa) [inline docs/queryworkers/akamai/index.md]: Great if we could link to the Akamai instructions for that.

File: docs/queryworkers/akamai/delete-edge-workers.md [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+---
+
+Delete an edge worker.
+
+1. [Log in to your Macrometa account](https://auth.paas.macrometa.io/).
+1. Click **Functions**.
+1. Click the stacked dots next to the edge worker you want to delete.
+1. Click **Delete Edge Worker**.
+1. Confirm your choice.
+    Macrometa permanently deletes the edge worker. You can re-create the edge worker, but you cannot undo the deletion.
```
Comment (diana-macrometa) [inline docs/queryworkers/akamai/delete-edge-workers.md]: Check your linter. Looks like there is one space too many in front of this paragraph.

File: docs/queryworkers/akamai/view-edge-workers.md (hunk 3 above)
Comment (diana-macrometa) [reply docs/queryworkers/akamai/view-edge-workers.md]: Probably should mention here that you can also view your edge workers in the list on Akamai.

**Review by diana-macrometa — CHANGES_REQUESTED** [review]
Great start. The tasks are very tight, great stuff.

File: docs/queryworkers/akamai/index.md (hunk 6 above)
Comment (ghost) [reply docs/queryworkers/akamai/index.md:7]: I'll defer to @dlozina-macrometa here as I'm still very fuzzy on these new capabilities 🙈

File: docs/queryworkers/akamai/generate-edge-workers.md (hunk 5 above)
Comment (jamesm-macrometa) [reply docs/queryworkers/akamai/generate-edge-workers.md:9]: I added a Stream Worker version of the topic, but unfortunately my phone started getting the internal error for the smoke environment so I can't actually check ☹️

**Comment by diana-macrometa** [comment]
> Do we want to include any additional concept info on Akamai and its integration with Macrometa? If so we change the current Index into a Task, and create a new Index that contains concept stuff.

Yes, at least a sentence or two. There's great content you can lift from the video. However, we MUST have this content merged today, so we can go back later and refine it.

> I'm still a bit unclear on the "edge worker" vs "edge function" terminology and which goes where.

Edge worker = Akamai's thing that we are integrating with as of 17.11
Edge function = Macrometa's thing, not yet implemented. Not sure what it looks like yet, used to be called "function worker."

File: docs/queryworkers/akamai/generate-edge-worker-stream.md [hunk 9]
```diff
@@ -0,0 +1,17 @@
+---
+sidebar_position: 25
+title: Generate Edge Worker from Stream Worker
+---
+
+Generate an edge worker from a Stream worker.
```
Comment (diana-macrometa) [inline docs/queryworkers/akamai/generate-edge-worker-stream.md]:
```suggestion
Generate an edge worker from a stream worker.
```

File: docs/queryworkers/akamai/generate-edge-worker-stream.md [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+1. [Log in to your Macrometa account](https://auth.paas.macrometa.io/).
+1. Click **Stream Workers**.
+1. Select a stream worker or [create a new one](../../cep/index.md).
```
Comment (diana-macrometa) [inline docs/queryworkers/akamai/generate-edge-worker-stream.md]: When we actually write a task for creating a stream worker, let's update this link.

File: docs/queryworkers/akamai/generate-edge-worker-stream.md [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+1. Click **Generate Edge Worker**.
+1. Enter the following details:
+    - **Name -** Choose a name for your edge worker.
+    - **Description -** (Optional) Add a description to specify what the edge worker does.
+    - **Function Worker Type -** Choose between **Stream Publisher** and **Stream Query**.
+    - **Stream Worker Name -** The stream worker this function is based on. Automatic.
+    - **Stream Name -** Add a name for the stream that will be created based on the old stream's code.
```
Comment (diana-macrometa) [inline docs/queryworkers/akamai/generate-edge-worker-stream.md]: The old stream? What?

File: docs/queryworkers/akamai/manage-integrations.md [hunk 12]
```diff
@@ -0,0 +1,43 @@
+---
+sidebar_position: 10
+title: Manage Integrations
+---
+
+This page describes how to integrate Akamai's edge worker functionality as a Macrometa functions.
+
+:::note
```
Comment (diana-macrometa) [inline docs/queryworkers/akamai/manage-integrations.md:8]: Please add this note after the intro of each topic in this folder.

File: docs/queryworkers/akamai/manage-integrations.md [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+This feature is currently in preview.
+:::
+
+## Prerequisites
+
+- A Macrometa account with sufficient permissions to view and edit query workers.
+- In Akamai, a Property created with your Macrometa GDN integration. Refer to [Create and configure a property](https://techdocs.akamai.com/api-definitions/docs/create-config-prop).
```
Comment (diana-macrometa) [inline docs/queryworkers/akamai/manage-integrations.md]:
```suggestion
- In Akamai, a Property created with your Macrometa GDN integration. For more information about Akamai Properties, refer to [Create and configure a property](https://techdocs.akamai.com/api-definitions/docs/create-config-prop).
```

**Review by diana-macrometa — APPROVED** [review]
Love it. A few minor things, plus we need a sentence or two about the feature. Ideas:

Akamai edge workers are a powerful compute engine for creating and deploying microservices at the edge. The combination of Akamai’s compute engine with the Macrometa GDN allows you to develop microservices with powerful data storage and streaming capabilities, all running milliseconds from your users.

Edge workers allow you to use JavaScript to preprocess data or perform other tasks related to Macrometa streams and stream workers.

----
I'll watch the video again, send you any other blurbs I can pick out.

**Comment by diana-macrometa** [comment]
From the demo video: 

Edge workers allow you to publish data to pipelines. You can continuously send data to an edge worker, which then can publish the data to a pipeline, a Macrometa stream, or both. 

Edge workers allow you to use JavaScript to preprocess or post-process data published to pipelines.

Edge workers allow you to query data inside streams and stream workers in real time. 

Sounds like the big selling points are:
- stateful processing (pre- or post-process data) with JavaScript
- publish to pipelines
- query real-time data in the pipeline, including data in motion (streams and stream workers)

-----
Hope that helps? I'm acting as a high-functioning parrot here, giving you what Durga said.

**Comment by jamesm-macrometa** [comment]
> From the demo video:
> 
> Edge workers allow you to publish data to pipelines. You can continuously send data to an edge worker, which then can publish the data to a pipeline, a Macrometa stream, or both.
> 
> Edge workers allow you to use JavaScript to preprocess or post-process data published to pipelines.
> 
> Edge workers allow you to query data inside streams and stream workers in real time.
> 
> Sounds like the big selling points are:
> 
>     * stateful processing (pre- or post-process data) with JavaScript
> 
>     * publish to pipelines
> 
>     * query real-time data in the pipeline, including data in motion (streams and stream workers)
> 
> 
> Hope that helps? I'm acting as a high-functioning parrot here, giving you what Durga said.

Thank you!  This is very helpful.  I'll incorporate this info into the concept topic.

File: docs/queryworkers/akamai/index.md (hunk 6 above)
Comment (ghost) [reply docs/queryworkers/akamai/index.md:7]: EdgeWorkers enables developers to create and deploy microservices across more than a quarter of a million edge servers deployed around the globe. ([Edge Workers](https://developer.akamai.com/akamai-edgeworkers-overview)) Akamai has a vast content delivery network that extends Macrometa's points of presence worldwide. We allow users to build stateful applications across Akamai's CDN. That is how I understood partnership, we extend Edge Workers in a way that they can now save the state. Does this make sense? @dallas-macrometa @diana-macrometa @jamesm-macrometa

File: docs/queryworkers/akamai/index.md (hunk 2 above)
Comment (jamesm-macrometa) [reply docs/queryworkers/akamai/index.md]: Confirmed with Dino that Edge Worker is capitalized in Akamai's GUI.

**Comment by diana-macrometa** [comment]
@jamesm-macrometa, I just noticed that Akamai EdgeWorker doesn't have the space between the last two words in their docs. Please update as necessary.

File: docs/queryworkers/akamai/index.md (hunk 6 above)
Comment (jamesm-macrometa) [reply docs/queryworkers/akamai/index.md:7]: Great stuff @dlozina-macrometa !  Thanks very much.

**Comment by jamesm-macrometa** [comment]
@diana-macrometa I updated the Index topic to include concept info, including the helpful tidbits Dino provided.  Can you take a look at your earliest convenience?  Thanks!

**Comment by jamesm-macrometa** [comment]
> @jamesm-macrometa, I just noticed that Akamai EdgeWorker doesn't have the space between the last two words in their docs. Please update as necessary.

Whoops, will fix that now.

File: docs/queryworkers/akamai/manage-integrations.md [hunk 14]
```diff
@@ -0,0 +1,43 @@
+---
+sidebar_position: 10
+title: Manage Integrations
+---
+
+This page describes how to integrate Akamai's Edge Worker functionality as a Macrometa functions.
```
Comment (diana-macrometa) [inline docs/queryworkers/akamai/manage-integrations.md]:
```suggestion
This page describes how to integrate Akamai's Edge Worker functionality with the Macrometa platform.
```

File: docs/queryworkers/akamai/manage-integrations.md [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+:::note
+This feature is currently in preview.
+:::
+
+## Prerequisites
+
+- A Macrometa account with sufficient permissions to view and edit query workers.
```
Comment (diana-macrometa) [inline docs/queryworkers/akamai/manage-integrations.md]:
```suggestion
- A Macrometa account with sufficient permissions to view and edit query workers or stream workers.
```

File: docs/queryworkers/akamai/index.md [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+
+:::note
+This feature is currently in preview.
+:::
+
+You can use Akamai Edge Workers to publish data to pipelines. You can continuously send data to an Edge Worker, which then can publish the data to a pipeline, a Macrometa stream, or both.
+
+Edge workers allow you to use JavaScript to preprocess or post-process data published to pipelines. Additionally, you can query data inside streams and stream workers in real time.
+
+For more information about Akamai Edge Workers, refer to the [Akamai Edge Workers overview](https://developer.akamai.com/akamai-edgeworkers-overview).
```
Comment (diana-macrometa) [inline docs/queryworkers/akamai/index.md]:
```suggestion
For more information about Akamai EdgeWorkers, refer to the [Akamai Edge Workers Overview](https://developer.akamai.com/akamai-edgeworkers-overview).
```

File: docs/queryworkers/akamai/manage-integrations.md [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+- In Akamai, a Property created with your Macrometa GDN integration. For more information about Akamai Properties, refer to [Create and configure a property](https://techdocs.akamai.com/api-definitions/docs/create-config-prop).
+
+## Add Integration
+
+Add Akamai metadata from your Macrometa GDN environment.
```
Comment (diana-macrometa) [inline docs/queryworkers/akamai/manage-integrations.md]:
```suggestion
Add Akamai metadata to your Macrometa GDN environment.
```

File: docs/queryworkers/akamai/manage-integrations.md [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+:::note
+This feature is currently in preview.
+:::
+
+## Prerequisites
+
+- A Macrometa account with sufficient permissions to view and edit query workers.
+- In Akamai, a Property created with your Macrometa GDN integration. For more information about Akamai Properties, refer to [Create and configure a property](https://techdocs.akamai.com/api-definitions/docs/create-config-prop).
+
+## Add Integration
```
Comment (diana-macrometa) [inline docs/queryworkers/akamai/manage-integrations.md:17]: After the new API is rolled out, let's see if there's an API way to do this and add a tab. Something for next week.

**Review by diana-macrometa — APPROVED** [review]
Looks good, just a few minor suggestions, plus the EdgeWorker bit that it looks like you might have finished as I was typing this.

Please make a Jira to track returning to this content to add API, SDK, and CLI tabs as they become available.

**Pull request merged**